core/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -363,8 +364,21 @@ private[chisel3] class ChiselContext() {
   val viewNamespace = Namespace.empty
 }
 
+/** Stores a [[Definition]] that is imported so that its Instances can be
+  * compiled separately.
+  */
+case class ImportedDefinitionAnnotation[T <: BaseModule with IsInstantiable](importedDefinition: Definition[T]) extends NoTargetAnnotation
+
 private[chisel3] class DynamicContext(val annotationSeq: AnnotationSeq, val throwOnFirstError: Boolean) {
+  val importedDefinitionsAnno = annotationSeq.collect{case a: ImportedDefinitionAnnotation[_] => a}
   val globalNamespace = Namespace.empty
+
+  // Ensure imported Definitions get the same name as their modules so that
+  // they can be properly linked to Instances that are separately compiled
```

Comment:
```suggestion
  // Ensure imported Definitions emit as ExtModules with the correct name
  // so that instantations will also use the correct name, and prevent any name conflicts with
  // Modules/Definitions in this elaboration
```